### PR TITLE
feat(mito): parquet memtable reader

### PR DESCRIPTION
**Pull request by v0y4g3r**
I hereby agree to the terms of the [GreptimeDB CLA](https://github.com/GreptimeTeam/.github/blob/main/CLA.md).

## What's changed and what's your intention?

This PR adds basic support for reading parquet files residing in memory, by reusing some structs and functions in sst mod.

Since the lack of keyword generics, we have to refactor the parquet reader by extracting the methods and functions that are not related to async/sync IO operations to a `[___]Base` structs that can be shared between memory and files.

## Checklist

- [x] I have written the necessary rustdoc comments.
- [x] I have added the necessary unit tests and integration tests.
- [ ] This PR requires documentation updates.

### CODE REVIEW

File: src/mito2/src/memtable/bulk/row_group_reader.rs [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+            .collect()
+    }
+
+    /// Creates a page reader to read column at `i`.
+    fn column_page_reader(&self, i: usize) -> parquet::errors::Result<Box<dyn PageReader>> {
+        if let Some(cached_pages) = &self.base.column_uncompressed_pages[i] {
+            debug_assert!(!cached_pages.row_group.is_empty());
+            // Hits the row group level page cache.
+            return Ok(Box::new(RowGroupCachedReader::new(&cached_pages.row_group)));
+        }
```
Comment (evenyag) [inline src/mito2/src/memtable/bulk/row_group_reader.rs:109]: Maybe we don't need to cache uncompressed pages for memtables.

Comment (v0y4g3r) [reply src/mito2/src/memtable/bulk/row_group_reader.rs]: yes. we can skip the check here. 0d0d6340f1

File: src/mito2/src/sst/parquet/reader.rs [hunk 2]
```diff
@@ -1118,10 +1119,56 @@ impl ParquetReader {
     }
 }
 
-/// Reader to read a row group of a parquet file.
-pub struct RowGroupReader {
+/// RowGroupReaderVirtual represents the fields that cannot be shared
+/// between different `RowGroupReader`s.
+pub(crate) trait RowGroupReaderVirtual: Send {
```
[the remark this answers is not in the document]
Comment (v0y4g3r) [reply src/mito2/src/sst/parquet/reader.rs]: 80b6f9d512

File: src/mito2/src/memtable/bulk/part.rs [hunk 3]
```diff
@@ -26,94 +28,229 @@ use datatypes::arrow::array::{
...
         }
 
-        dest.metadata = BulkPartMeta {
-            num_rows: arrow_record_batch.num_rows(),
-            max_timestamp: max_ts,
-            min_timestamp: min_ts,
-        };
-        Ok(true)
+        let buf = Bytes::from(buf);
+        let parquet_metadata = load_metadata(&buf)?;
```
[the remark this answers is not in the document]
Comment (v0y4g3r) [reply src/mito2/src/memtable/bulk/part.rs]: > I remember that we have a helper function to convert the metadata to parquet metadata.

sound. fixed in d4d6d3c58e

File: src/mito2/src/memtable/bulk/part.rs [hunk 4]
```diff
@@ -26,94 +28,229 @@ use datatypes::arrow::array::{
...
+    row_group_size: Option<usize>,
     dedup: bool,
+    writer_props: Option<WriterProperties>,
+}
+
+impl BulkPartEncoder {
+    pub(crate) fn new(
+        metadata: RegionMetadataRef,
+        dedup: bool,
+        row_group_size: Option<usize>,
```
[the remark this answers is not in the document]
Comment (v0y4g3r) [reply src/mito2/src/memtable/bulk/part.rs]: fixed in bb57cd2a921ee52e21692e19e64b2fd3a1148802

File: src/mito2/src/memtable/bulk/part.rs [hunk 5]
```diff
@@ -26,94 +28,229 @@ use datatypes::arrow::array::{
...
+        let iter = BulkPartIter::try_new(
+            context,
+            row_groups_to_read,
+            self.metadata.parquet_metadata.clone(),
+            self.data.clone(),
+        )?;
+        Ok(Some(Box::new(iter) as BoxedBatchIterator))
+    }
+
+    fn build_read_format(&self, projection: &Option<&[ColumnId]>) -> ReadFormat {
```
[the remark this answers is not in the document]
Comment (v0y4g3r) [reply src/mito2/src/memtable/bulk/part.rs]: > If we have multiple parts, then they should reuse the same `ReadFormat`. Maybe we shouldn't define it as a method of the part.

fixed in ae32bc5496

File: src/mito2/src/sst/parquet/stats.rs [hunk 6]
```diff
@@ -27,7 +27,7 @@ use store_api::storage::ColumnId;
 use crate::sst::parquet::format::ReadFormat;
 
 /// Statistics for pruning row groups.
-pub(crate) struct RowGroupPruningStats<'a, T> {
+pub struct RowGroupPruningStats<'a, T> {
```
Comment (evenyag) [inline src/mito2/src/sst/parquet/stats.rs:30]: Do we need to pub this? Maybe `pub(crate)` is enough?

File: src/mito2/src/sst/parquet/row_group.rs [hunk 7]
```diff
@@ -38,25 +38,194 @@ use crate::sst::file::FileId;
...
+            })
+            .map(|(idx, (_chunk, _pages))| {
+                let column = self.metadata.column(idx);
+                let (start, length) = column.byte_range();
+                start..(start + length)
+            })
+            .collect::<Vec<_>>()
+    }
+
+    pub(crate) fn assign_dense_chunk(
```
Comment (evenyag) [inline src/mito2/src/sst/parquet/row_group.rs:159]: nitpick: Maybe add some documents to the input and return value of some methods in `RowGroupBase`.

**Review by evenyag — APPROVED** [review]
Others LGTM

**Review by waynexia — APPROVED** [review]

**Comment by fengjiachun** [comment]
Some conflits, PTAL @v0y4g3r

**Pull request merged**